irohad/main/impl/pg_connection_init.cpp

```diff
@@ -5,16 +5,164 @@
 
 #include "main/impl/pg_connection_init.hpp"
 
+#include <array>
+#include <random>
+
+#include <boost/functional/hash.hpp>
+#include <boost/range/adaptor/transformed.hpp>
+#include "common/irohad_version.hpp"
 #include "logger/logger.hpp"
 #include "logger/logger_manager.hpp"
 
+// disabling GNU macros
+#ifdef major
+#undef major
+#endif
+
+#ifdef minor
+#undef minor
+#endif
+
+using namespace iroha::ametsuchi;
+
 namespace {
+  using iroha::IrohaVersionType;
+  constexpr IrohaVersionType kSchemaVersionMajor = 0;
+  constexpr IrohaVersionType kSchemaVersionMinor = 0;
+
+  /**
+   * This structure represents a database schema version.
+   *
+   * The logic behind schema versioning is the following:
+   *  - user may try to launch an iroha binary of any version
+   *  - user may have a database created and previously used by an iroha binary
+   *    of any version
+   *  - versions of iroha are organized in form of a tree: any new version has
+   *    only one previous version that itself can not be parent of the former
+   *    one.
+   *  - supported iroha versions are uniquely identified by 3 numbers: major,
+   *    minor and patch
+   *  - some different supported iroha versions may, and some require to be
+   *    compatible with some other versions
+   *  - over time, different irohad source and binaries can be tagged with same
+   *    supported version numbers, but the latter will be always children of the
```

Comment:
"latter" or "later"?
Also, "children" assumes direct "parent - child" relationship between two versions.
Maybe something like this:

> - if two different irohad sources and binaries are tagged with the same supported version number then the newer ones must be descendants of the older